File: 2018/Day08.py
def read(numbers):
    children, metadata = next(numbers), next(numbers)
    return ([read(numbers) for _ in range(children)],
            [next(numbers) for _ in range(metadata)])

def total(node):
    children, metadata = node
    return sum(metadata) + sum(total(c) for c in children)

def value(node):
    children, metadata = node
    if not children:
        return sum(metadata)
    return sum(value(children[m-1]) for m in metadata if 1 <= m <= len(children))

def part_1(input):
    return total(input)

def part_2(input):
    return value(input)

with open("2018/Day08_input.txt") as f:
    input = read(iter([int(x) for x in f.read().split()]))

    print(f"Part 1: {part_1(input)}")
    print(f"Part 2: {part_2(input)}")

File: 2022/Day08.go
package main

import (
	"fmt"
	"os"
	"strings"
)

var dirs = [4][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

func look(g []string, i, j, di, dj int) (int, bool) {
	d := 0
	for y, x := i+di, j+dj; y >= 0 && y < len(g) && x >= 0 && x < len(g[y]); y, x = y+di, x+dj {
		d++
		if g[y][x] >= g[i][j] {
			return d, false
		}
	}
	return d, true
}

func part1(in string) int {
	g, n := strings.Fields(in), 0
	for i := range g {
		for j := range g[i] {
			for _, d := range dirs {
				if _, edge := look(g, i, j, d[0], d[1]); edge {
					n++
					break
				}
			}
		}
	}
	return n
}

func part2(in string) int {
	g, best := strings.Fields(in), 0
	for i := range g {
		for j := range g[i] {
			s := 1
			for _, d := range dirs {
				v, _ := look(g, i, j, d[0], d[1])
				s *= v
			}
			if s > best {
				best = s
			}
		}
	}
	return best
}

func main() {
	b, _ := os.ReadFile("2022/Day08_input.txt")
	in := strings.TrimRight(string(b), "\n")
	fmt.Println("Part 1:", part1(in))
	fmt.Println("Part 2:", part2(in))
}
